mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOpsTypes.h

```diff
@@ -41,567 +42,34 @@ enum class MemoryTargetMode {
-
-  bool operator==(const KeyTy &Key) const {
-    return Key == KeyTy{Type, Dimension, AccessMode, TargetMode, Body};
-  }
-
-  static llvm::hash_code hashKey(const KeyTy &Key) {
-    return llvm::hash_combine(std::get<0>(Key), std::get<1>(Key),
-                              std::get<2>(Key), std::get<3>(Key),
-                              std::get<4>(Key));
-  }
-
-  static KeyTy getKey(const KeyTy &Key) { return KeyTy{Key}; }
-
-  static AccessorTypeStorage *construct(TypeStorageAllocator &Allocator,
-                                        const KeyTy &Key) {
-    return new (Allocator.allocate<AccessorTypeStorage>())
-        AccessorTypeStorage(Key);
-  }
-
-  mlir::Type Type;
-  unsigned int Dimension;
-  mlir::sycl::MemoryAccessMode AccessMode;
-  mlir::sycl::MemoryTargetMode TargetMode;
-  llvm::SmallVector<mlir::Type, 4> Body;
-};
-
-struct RangeTypeStorage : public TypeStorage {
-  using KeyTy = unsigned int;
-
-  RangeTypeStorage(const KeyTy &Key) : Dimension(Key) {}
-
-  bool operator==(const KeyTy &Key) const { return Key == KeyTy{Dimension}; }
-
-  static llvm::hash_code hashKey(const KeyTy &Key) {
-    return llvm::hash_value(Key);
-  }
-
-  static KeyTy getKey(const KeyTy &Key) { return KeyTy{Key}; }
-
-  static RangeTypeStorage *construct(TypeStorageAllocator &Allocator,
-                                     const KeyTy &Key) {
-    return new (Allocator.allocate<RangeTypeStorage>()) RangeTypeStorage(Key);
-  }
-
-  unsigned int Dimension;
-};
-
-struct NdRangeTypeStorage : public TypeStorage {
-  using KeyTy = std::tuple<unsigned int, llvm::SmallVector<mlir::Type, 4>>;
-
-  NdRangeTypeStorage(const KeyTy &Key)
-      : Dimension(std::get<0>(Key)), Body(std::get<1>(Key)) {}
-
-  bool operator==(const KeyTy &Key) const {
-    return Key == KeyTy{Dimension, Body};
-  }
-
-  static llvm::hash_code hashKey(const KeyTy &Key) {
-    return llvm::hash_combine(std::get<0>(Key), std::get<1>(Key));
-  }
-
-  static KeyTy getKey(const KeyTy &Key) { return KeyTy{Key}; }
-
-  static NdRangeTypeStorage *construct(TypeStorageAllocator &Allocator,
-                                       const KeyTy &Key) {
-    return new (Allocator.allocate<NdRangeTypeStorage>())
-        NdRangeTypeStorage(Key);
-  }
-
-  unsigned int Dimension;
-  llvm::SmallVector<mlir::Type, 4> Body;
-};
-
-struct AccessorImplDeviceStorage : public TypeStorage {
-  using KeyTy = std::tuple<unsigned int, llvm::SmallVector<mlir::Type, 4>>;
-
-  AccessorImplDeviceStorage(const KeyTy &Key)
-      : Dimension(std::get<0>(Key)), Body(std::get<1>(Key)) {}
 
-  bool operator==(const KeyTy &Key) const {
-    return Key == KeyTy{Dimension, Body};
-  }
+llvm::StringRef
+memoryAccessModeAsString(mlir::sycl::MemoryAccessMode MemAccessMode);
+LogicalResult parseMemoryAccessMode(AsmParser &Parser,
+                                    FailureOr<MemoryAccessMode> &MemAccessMode);
+void printMemoryAccessMode(AsmPrinter &Printer, MemoryAccessMode MemAccessMode);
```

Comment:
`<X>AsString` is a utility function used by `print<X>` and to obtain aliases of types.
`parse<X>` and `print<X>` are required by the assembly format defined in the `.td` file.